Introduce support of nmpolicy

cathay4t opened the pull request:
Introducing support of nmpolicy syntax[1] including rust, python and C
bindings.

Implementation workflow:
 * `NetworkPolicy` could deserialize from existing nmpolicy file.
 * Each of capture line will produce a new `NetworkState` and stored for
   follow up capture line processing.
 * Each capture line will be converted to `Vec<NetworkCaptureToken>`,
   then fill into `NetworkCaptureCommand` for filtering or updating
   matched `NetworkState`.
 * We use `serde_json::Map` to walk through the property defined in the
   format of `routes.running.next-hop-interface`.

The cli tools could be used also:

    nmstatectl p <policy.yml> [-c current_state.yml] \
        [-s saved_capture.yml] [-o /path/to/output_capture.yml]


API:

 * Using running current to generate `NetworkState`:

   ```rust
   let policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   let state = NetworkState::try_from(policy)?
   ```

 * Use saved current to generate `NetworkState`:

   ```rust
   let mut policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   policy.current = Some(serde_yaml::from_str(current_state_content)?);
   let state = NetworkState::try_from(policy)?
   ```

 * Store captured data:

   ```rust
   let policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   let current_state = serde_yaml::from_str(current_state_content)?;
   let captured_states = policy.capture.execute(&current_state)?;
   // Use your own way to store the captured data
   // You may also use `NetworkCaptureRules::execute()` directly.
   ```

 * Load captured data:

   ```rust
   // Use your own way to load the data. We just need
   // HashMap<String, NetworkState> for captured_states
   let current_state = serde_yaml::from_str(current_state_content)?;
   let policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   let state = policy.desired.fill_with_captured_data(&captured_states)?;
   // You may also use `NetworkStateTemplate::fill_with_captured_data()`
   // directly.
   ```

Both unit test cases and integration test case are included.

qinqon commented on rust/src/cli/ncl.rs:
```diff
@@ -255,6 +260,30 @@ fn main() {
+                        .takes_value(false)
+                        .help("Show state in json format"),
+                )
+        )
```
we need also a captured state cache arguments one for input and another for output.

https://github.com/nmstate/nmpolicy/blob/main/cmd/nmpolicyctl/gen.go#L82-L86

cathay4t replied:
Why we need that? Any use case require so?

qinqon replied:
The main use case is policy re-apply (this happends at kubernetes-nmstate if the handle pod is restarted), it has to generate the same state and for that you need to pass the previously captured values, if you don't do so a capture entry like `routes.destination=="0.0.0.0/0"` will return different value. It's part of the design doc of nmpolicy too.

This is some integration test from nmpolicy golang https://github.com/nmstate/nmpolicy/blob/main/tests/nmstatectl_test.go#L96

cathay4t replied:
The capture state is generated by policy and current state. If you provide identical current state, you get identical intermediate capture states. Hence I don't know why we need to expose this intermediate state.

qinqon replied:
For kubernetes-nmstate saving the previous network state per node per NNCP does not scale well, that's why se just save per capture.

qinqon commented on rust/src/cli/ncl.rs:
```diff
@@ -255,6 +260,30 @@ fn main() {
+                        .takes_value(true)
+                        .help("Read current network state from file"),
+                )
+                .arg(
```
Can we add another one to also apply it ? `--apply` or even the opposite apply by default and just dump if `--no-apply` is passed.

cathay4t replied:
In V3 patch, the `nmstatectl policy` just for generate yaml. The `nmstatectl apply` support both policy and state with the same other options.

qinqon commented on rust/src/clib/test/nmpolicy_test.c:
```diff
@@ -0,0 +1,99 @@
+int main(void) {
+	int rc = EXIT_SUCCESS;
+	const char *policy = "{"
+"  \"capture\": {"
+"    \"default-gw\": \"override me with the cache\","
+"    \"base-iface\": \"interfaces.name== capture.default-gw.routes.running.0.next-hop-interface\","
+"    \"base-iface-routes\": \"routes.running.next-hop-interface== capture.default-gw.routes.running.0.next-hop-interface\","
+"    \"bridge-routes\": \"capture.base-iface-routes | routes.running.next-hop-interface:=\\\"br1\\\"\""
+"  },"
+"  \"desiredState\": {"
+"    \"interfaces\": ["
+"      {"
+"        \"name\": \"br1\","
+"        \"description\": \"Linux bridge with base interface as a port\","
+"        \"type\": \"linux-bridge\","
+"        \"state\": \"up\","
+"        \"ipv4\": \"{{ capture.base-iface.interfaces.0.ipv4 }}\","
+"        \"bridge\": {"
+"          \"options\": {"
+"            \"stp\": {"
+"              \"enabled\": false"
+"            }"
+"          },"
+"          \"port\": ["
+"            {"
+"              \"name\": \"{{ capture.base-iface.interfaces.0.name }}\""
+"            }"
+"          ]"
+"        }"
+"      }"
+"    ],"
+"    \"routes\": {"
+"      \"config\": \"{{ capture.bridge-routes.routes.running }}\""
+"    }"
+"  }"
+"}";
```
We should have the nmpolicy integration tests at files so they can also be used by the bindings.

cathay4t replied:
We already have

```
dynamic_ip_test.py::test_policy_from_auto_nic_to_bridge
route_test.py::test_policy_from_static_eth1_to_bridge
```

This unit test in C binding is just proving the C wrapping is working, not the real functions.
The rust unit test and python integration is testing the real functionality.

qinqon commented on tests/integration/route_test.py:
```diff
@@ -1209,3 +1212,71 @@ def test_sanitize_route_rule_from_to(route_rule_test_env):
+
+
+@pytest.mark.tier1
+def test_policy_from_static_eth1_to_bridge(eth1_static_gateway_dns):
```
We should pass all the test under `https://github.com/nmstate/nmpolicy/tree/main/tests` also it would better if they follow the same file structure for test data.

cathay4t replied:
V3 patch ported these files and unit test is passing.

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+impl TryFrom<String> for NetworkCaptureCondition {
+    type Error = NmstateError;
+    fn try_from(s: String) -> Result<NetworkCaptureCondition, NmstateError> {
+        let action = if s.contains(&NetworkCaptureAction::Equal.to_string()) {
```
You can have something like foo.bar := "foo==bar" and it will think that is an "filter" operation when is a "replace", that's why is better to tokenize the command first, so first you chop the string propertly then you iterate the chops.

The proposed tokenizer at nmpolicy would be good enough, https://github.com/nmstate/nmpolicy/tree/rust/rust/src/lib/lex

cathay4t replied:
Fixed by new patch using `NetworkCaptureToken`

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct NetworkCaptureRules {
```
These are not rules, Just NetworkCapture is enough it contains all the entries.

qinqon replied:
This should be a map and I think "Rule" does not fit her we can just use  `NetworkCapture`

```rust
pub struct NetworkCapture {
    pub capture: HashMap<String, NetworkCaptureCommand>
}
```

cathay4t replied:
HashMap does not preserve the order which matters for generating capture states. Hence I use Vec.

For naming, I prefer it as it is.

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+    ) -> Result<HashMap<String, NetworkState>, NmstateError> {
+        let mut ret = HashMap::new();
+        for (var_name, condition) in self.rules.as_slice() {
+            let matched_state = condition.get_match(current, &ret)?;
```
The name is not right, we have commands that "filter" and command that "replace" this would be something like "command.execute(current, &ref)?;"

cathay4t replied:
Changed to `execute()`.

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+}
+
+impl NetworkCaptureCondition {
+    fn get_match(
```
As say better use the terms `NetworkCaptureCommand` and `execute` instead of `get_match`

cathay4t replied:
Changed to `NetworkCaptureCommand`.

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+        let matching_value =
+            value_to_string(&get_value(self.value.as_slice(), &ret, captures)?);
+
+        match self.key.get(0).map(String::as_str) {
```
At nmpolicy we don't restrict to routes or interfaces so users can "filter" or "replace" any of the top level fields at nmstate.

cathay4t replied:
For filter and replace, they are working against array of items. The interface itself is a array, but for route, route rule, they are under `running` or `config`.

To support any field match, we need more code to confirm user you working against a array. Considering we only have `interfaces`, `route`, `route-rules`, I like current workflow.

For filter and replace on `dns-resolver`, there is no use of it, user can just specific them in desire state.

The V2 patch support `retain_only` action for `dns-resolver` like `dns: dns-resolver.running`, but I don't see a point why it is require, use can just refer it directly in the template.

qinqon replied:
At nmpolicy to generalize we just filter at the first slice that appears but not at the maps, doing so we can generalize independently of the deep of the struct.

For replace is kind of the same but instead of filtering the slice we just replace one value with another on the first slice.

Generally speaking we should support only the scenarios that we test, but if the implementation is generic enough we don't stop users from experiment with not tested support (if it's worth it can be tested and supported later on).

cathay4t replied:
I don't like write code for use case base on my guess.

I like to quote the PEP 20 -- Zen of Python:
```
Explicit is better than implicit. Simple is better than complex. 
Complex is better than complicated. Flat is better than nested.
```

qinqon replied:
Ok, let's start with this we can generalize later on if needed.

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+        let splits: Vec<&str> = s.split(&action.to_string()).collect();
+        if splits.len() == 2 {
+            Ok(NetworkCaptureCondition {
+                capture_name: get_capture_name_if_piped(splits[0])?,
```
We should handle Pipes like a concatenation of commands not just as syntatic suggar for an argument of a command.

I understand that we want to limit what users can do with expresions but let' not put that limites when we parse the strings.

cathay4t replied:
V2 patch updated for handling this key capture or value capture.

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+            Ok(NetworkCaptureCondition {
+                capture_name: get_capture_name_if_piped(splits[0])?,
+                key: parse_string_to_property_list(splits[0]),
+                value: parse_string_to_property_list(splits[1]),
```
We should use a Rust Enum to differenciate between different types of value, "string", number or path

cathay4t replied:
Unless we see problem break by this, I intend to keep it as it is.

This AST stuff could be easily goes crazy if I loosen the reins. I am not building another language over yaml, all code change should be use case based.

qinqon replied:
Fair enough

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,367 @@
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(deny_unknown_fields, try_from = "String", into = "String")]
+#[non_exhaustive]
+pub struct NetworkCaptureCondition {
```
More than `Condition` this is `Command` and instead of this kind of ad-hoc struct is better represented as as some traits, enums and structs to as a tree.

Something in the line of:

```rust
use std::fmt;
use std::rc::Rc;

#[derive(Debug)]
pub struct NetworkState {
    value: String,
}
pub struct NmstateError {}

pub trait Command {
    fn execute(&self) -> Result<NetworkState, NmstateError>;
}

#[derive(Debug)]
pub enum Step {
    String(String),
    Number(i32),
}

#[derive(Debug)]
pub struct Path {
    steps: Vec<Step>,
}

impl fmt::Display for Path {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "path: {:?}", self.steps)
    }
}

#[derive(Debug)]
pub enum Value {
    String(String),
    Number(i32),
    Boolean(bool),
    Path(Path),
}

pub struct FilterCommand {
    input: Box<dyn Command>,
    path: Path,
    value: Value,
}

pub struct ReplaceCommand {
    input: Box<dyn Command>,
    path: Path,
    value: Value,
}

pub struct PathCommand {
    input: Box<dyn Command>,
    path: Path,
}

pub struct ExecuteCaptureCommand {
    name: String,
    capture: Rv<Map<String, Box<dyn Command>>>,
}

pub struct ShowCommand {}

impl Command for FilterCommand {
    fn execute(&self) -> Result<NetworkState, NmstateError> {
        Ok(NetworkState {
            value: format!(
                "filter.execute input: {:?}, path: {:?}, value: {:?}",
                self.input.execute()?,
                self.path,
                self.value
            ),
        })
    }
}
impl Command for ReplaceCommand {
    fn execute(&self) -> Result<NetworkState, NmstateError> {
        Ok(NetworkState {
            value: format!(
                "replace.execute input: {:?}, path: {:?}, value: {:?}",
                self.input.execute()?,
                self.path,
                self.value
            ),
        })
    }
}

impl Command for PathCommand {
    fn execute(&self) -> Result<NetworkState, NmstateError> {
        Ok(NetworkState {
            value: format!(
                "path.execute input: {:?}, path: {:?}",
                self.input.execute()?,
                self.path
            ),
        })
    }
}

impl Command for ShowCommand {
    fn execute(&self) -> Result<NetworkState, NmstateError> {
        println!("show.execute");
        Ok(NetworkState {
            value: "currentState".to_string(),
        })
    }
}

impl Command for ExecuteCommand {
    fn execute(&self) -> Result<NetworkState, NmstateError> {}
}

fn main() {
    println!("Examples from http://nmstate.io/nmpolicy/examples.html");
    println!("routes.running.destination=='0.0.0.0/0'");
    let filter = FilterCommand {
        input: Box::new(ShowCommand {}),
        path: Path {
            steps: vec![
                Step::String("routes".to_string()),
                Step::String("running".to_string()),
                Step::String("destination".to_string()),
            ],
        },
        value: Value::String("0.0.0.0/0".to_string()),
    };
    if let Ok(captured_state) = filter.execute() {
        println!("{}", captured_state.value);
    }

    println!("interfaces.name==capture.default-gw.routes.running.0.next-hop-interface");
    let filter = FilterCommand {
        input: Box::new(ShowCommand {}),
        path: Path {
            steps: vec![
                Step::String("interfaces".to_string()),
                Step::String("name".to_string()),
            ],
        },
        value: Value::Path(Path {
            steps: vec![
                Step::String("capture".to_string()),
                Step::String("default-gw".to_string()),
                Step::String("routes".to_string()),
                Step::String("running".to_string()),
                Step::Number(0),
                Step::String("next-hop-interface".to_string()),
            ],
        }),
    };
    if let Ok(captured_state) = filter.execute() {
        println!("{}", captured_state.value);
    }

    println!("capture.linux-bridges | interfaces.state:='down'");
}


```

Then first you compose the Tree of commands if a PathCommand is refering a capture you will have to execute the capture command first and cache the value and continue so it has the value.

cathay4t replied:
V2 patch improved this work flow. Now is:

`Sting -> Vec<Token> -> Command -> CaptureRules -> captured NetworkState.`

qinqon reviewed (changes requested):
First pass review focusing on the syntax parsing the main concerns:
- We should tokenize and convert this to tree (it does not have to be an AST)
- About naming let's use "commands" concept instead of "conditions" 
- We don't need to limit ourselves to "interfaces" and "routes" if we generalize a little the that navigate the json map.
- Error handling has to be near equal to golang implementation the main missing part is the snippet of expression that show you where in the expression this is failing, there is a rust implementation here https://github.com/nmstate/nmpolicy/blob/rust/rust/src/lib/snippet.rs, for that we need to store the string position over tokens/tree so when we are dealing with them we can access and return errors

cathay4t commented:
V2 patch uploaded with `NetworkCaptureToken` doing the AST stuff in the simplest way.

I am adding more the test cases.

cathay4t commented:
V3 patch uploaded with test files copy from https://github.com/nmstate/nmpolicy/tree/main/docs/examples [1]

And also `nmstatectl apply` support both state and policy file. The `nmstatectl policy` is just for generating.

[1]: Only fixed the incorrect indent in `bridge-on-default-gw-dhcp` which `port` should not been placed under bridge `options`. Of course, this found by nmstate deserilizer.

cathay4t commented:
/retest

qinqon commented on rust/src/lib/policy/token.rs:
```diff
@@ -0,0 +1,150 @@
+const NEXT_TOKEN_START_CHARS: [char; 6] = [' ', '=', '!', '|', '"', ':'];
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub(crate) enum NetworkCaptureToken {
```
Path is too big for a token and Value can be more concrete, we can do something like 

```rust
pub(crate) enum Token {
    Identity(usize, String), // foo
    Number(usize, i32), // 23
    Str(usize, String), // "foo"
    Dot(usize),      // .
    Pipe(usize),     // |
    Replace(usize),  // !=
    Filter(usize), // ==
}
```

Then we construct Path later on as a vector of enums as part of the commands tree.

cathay4t replied:
If we do that, we are introducing complexity for no reason. Why twice?

qinqon replied:
Ok at least lest identify at tokenize between String, Number and Identity.

qinqon commented on examples/policy/all-ethernet-up/current.yml:
```diff
@@ -0,0 +1,26 @@
+---
+interfaces:
```
Does it make sense to get the yamls from "github.com/nmstate/nmpolicy" ? at least shile nmpolicy is not from nmstatectl everywhere, this way if we add something at nmpolicy we know is suported at the nmstatectl version.

cathay4t replied:
Once this PR is merged. The `github.com/nmstate/nmpolicy` should be in archive state.

qinqon commented on rust/src/lib/policy/capture.rs:
```diff
@@ -0,0 +1,528 @@
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(deny_unknown_fields, try_from = "String", into = "String")]
+#[non_exhaustive]
+pub struct NetworkCaptureCommand {
```
Let's have specific commands so they have specific fields and we don't mix the fact that a capture is piped there


```rust
pub trait Command {                                                             
    fn execute(&self) -> Result<NetworkState, NmstateError>;                    
}    

pub struct FilterCommand {                                                      
    input: Box<dyn Command>,                                                    
    path: Path,                                                                 
    value: Value,                                                               
}                                                                               
                                                                                
pub struct ReplaceCommand {                                                     
    input: Box<dyn Command>,                                                    
    path: Path,                                                                 
    value: Value,                                                               
}                                                                               
                                                                                
pub struct PathCommand {                                                        
    input: Box<dyn Command>,                                                    
    path: Path,                                                                 
}                                                                               
                                                                                
pub struct ExecuteCaptureCommand {                                              
    name: String,                                                               
    capture: Rv<Map<String, Box<dyn Command>>>,                                 
}                                                              
```

We can even use memoization and cache the Command::Execute result in the future.

cathay4t replied:
I don't know what this future looks like and why current code cause problem. Can you elaborate?

qinqon replied:
Fair enough, I think we can refactor later on.

qinqon reviewed (changes requested):
Another high level review, we are still missing the error handling with the snippet.

Also can we do the c binding at different PR or later on so we reduce the size of this one ?

cathay4t commented:
> Another high level review, we are still missing the error handling with the snippet.

There is error handling in the code, which part are you expecting?

> 
> Also can we do the c binding at different PR or later on so we reduce the size of this one ?

The binding is required for CI test case. It is required for nmstate PR to have test approving it is working as described.

qinqon commented:
> > Another high level review, we are still missing the error handling with the snippet.
> 
> There is error handling in the code, which part are you expecting?

The error should show the snippet of the capture entry and a pointer where the error is in the expression

Here is some integration tests that we have to also support 

https://github.com/nmstate/nmpolicy/blob/rust/tests/nmstatectl_test.go#L146

cathay4t commented:
Coding V4 patch for change requests:
 * Support `NetworkStateTemplate::fill_with_captured_data()` and CLI.
 * Support `NetworkPolicy::gen_capture_states()` and CLI.
 * Use `ErrorKind::PolicySyntaxError` with position information and CLI render it with code snippet.

cathay4t commented:
V4 patch uploaded.

 * Introduce `NetworkCaptureRules::execute()` to stored captured states.
 * Introduce `NetworkStateTemplate::fill_with_captured_data()` to load captured states.
 * The old capture schema in nmpolicy-go is not supported at API level. CLI support it for dumping and loading.
 * Support positioning the error context, the `Display` trait of `NmstateError` will provide handy outputs. example:

```
PolicyError: Invalid pipe action: no property path defined
| capture.default-gw |
| ...................^
```
 * A lot unit test case included to ensure we are positioning the syntax error. Looks over-kill to me.
 * The patch size goes to 3000+. So I would like to limit the review on basic issue this PR not solving, we can do the small improvement later in other small PRs.

qinqon reviewed (changes requested):
Found some functional issues running the PR at nmstate towards nmpolicy project integration tests.

**From nmpolicy project examples integration tests**


For the example that [put all the linux bridges down](http://nmstate.io/nmpolicy/examples.html#set-all-linux-bridges-down), looks like at the generated state it's dumping empty structs for fields not modified, nmstatectl policy should return the minimal state to change what is supose to change, in this case only `interfaces`.
```
Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,4 @@
            	            	-(map[string]interface {}) (len=1) {
            	            	+(map[string]interface {}) (len=4) {
            	            	+ (string) (len=12) "dns-resolver": (map[string]interface {}) {
            	            	+ },
            	            	  (string) (len=10) "interfaces": ([]interface {}) (len=3) {
            	            	@@ -17,2 +19,6 @@
            	            	   }
            	            	+ },
            	            	+ (string) (len=11) "route-rules": (map[string]interface {}) {
            	            	+ },
            	            	+ (string) (len=6) "routes": (map[string]interface {}) {
            	            	  }

```

**From nmpolicy project nmpolicyctl integration tests**

- [Should dump empty state if the policy is empty](https://github.com/nmstate/nmpolicy/blob/main/tests/nmstatectl_test.go#L53).
- [Should not fail if there is no "capture" field it will just dump desiredState (this is important for knmstate)](https://github.com/nmstate/nmpolicy/blob/main/tests/nmstatectl_test.go#L63).
- Errors are pointing to a missleading character (description although different is good enough): 
[test](https://github.com/nmstate/nmpolicy/blob/main/tests/nmstatectl_test.go#L146)
```
expected: 'Error: failed to generate state, err: failed to resolve capture expression, err: illegal rune -
| routes.running.destination==-\"0.0.0.0/0\"                                  
| ............................^                                               
': exit status 1"                                                               
actual:'Failed to load from file, tried both YAML and JSON format. Errors: PolicyError: The equal or replace action should end with single value or property path
| routes.running.destination==-\"0.0.0.0/0\"                                  
| ..............................^, expected value at line 1 column 1          
```

[test](https://github.com/nmstate/nmpolicy/blob/main/tests/nmstatectl_test.go#L174)
```
expected: 'Error: failed to generate state, err: failed to resolve capture expression, err: resolve error: eqfilter error: invalid path input source (Path=[Identity=interfaces]), only capture reference is supported
| interfaces | routes.running.destination==\"0.0.0.0/0\"                      
| ^                                                                           
': exit status 1"                                                               
actual  : 'Failed to load from file, tried both YAML and JSON format. Errors: PolicyError: The pipe action should always in format of 'capture.<capture_name>'
| interfaces | routes.running.destination==\"0.0.0.0/0\"                      
| ...........^, expected value at line 1 column 1                             
': exit status 1"                                       
```

cathay4t commented:
Thanks for the review.

V5 patch uploaded with fixes of https://github.com/nmstate/nmstate/pull/2050#pullrequestreview-1152819226

qinqon reviewed (changes requested):
When the generated state is empty is should be `{}` but looks like it's marshaled as `{"interfaces":{}, "routes":{}}`

cathay4t commented:
> When the generated state is empty is should be `{}` but looks like it's marshaled as `{"interfaces":{}, "routes":{}}`

That is empty `NetworkState`. Any problem caused by that?

cathay4t commented:
> 

Never mind. Changed empty policy got empty string return.

cathay4t committed:
Introduce support of nmpolicy

Introducing support of nmpolicy syntax[1] including rust, python and C
bindings.

Implementation workflow:
 * `NetworkPolicy` could deserialize from existing nmpolicy file.
 * Each of capture line will produce a new `NetworkState` and stored for
   follow up capture line processing.
 * Each capture line will be converted to `Vec<NetworkCaptureToken>`,
   then fill into `NetworkCaptureCommand` for filtering or updating
   matched `NetworkState`.
 * We use `serde_json::Map` to walk through the property defined in the
   format of `routes.running.next-hop-interface`.

The cli tools could be used also:

    nmstatectl p <policy.yml> [-c current_state.yml] \
        [-s saved_capture.yml] [-o /path/to/output_capture.yml]

API:

 * Using running current to generate `NetworkState`:

   ```rust
   let policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   let state = NetworkState::try_from(policy)?
   ```

 * Use saved current to generate `NetworkState`:

   ```rust
   let mut policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   policy.current = Some(serde_yaml::from_str(current_state_content)?);
   let state = NetworkState::try_from(policy)?
   ```

 * Store captured data:

   ```rust
   let policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   let current_state = serde_yaml::from_str(current_state_content)?;
   let captured_states = policy.capture.execute(&current_state)?;
   // Use your own way to store the captured data
   // You may also use `NetworkCaptureRules::execute()` directly.
   ```

 * Load captured data:

   ```rust
   // Use your own way to load the data. We just need
   // HashMap<String, NetworkState> for captured_states
   let current_state = serde_yaml::from_str(current_state_content)?;
   let policy: NetworkPolicy = serde_yaml::from_str(policy_content)?;
   let state = policy.desired.fill_with_captured_data(&captured_states)?;
   // You may also use `NetworkStateTemplate::fill_with_captured_data()`
   // directly.
   ```

Both unit test cases and integration test case are included.

[1]: https://nmstate.io/nmpolicy/user-guide/102-policy-syntax.html

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
Got go-sign from @qinqon
